### PR TITLE
FileSystem: Implement WinRT File Enumeration

### LINKED ISSUE #8108: Implement File enumeration over WinRT
We are missing an implementation of FileSystem.EnumeratePaths and FileSystem.EnumerateFileSystemInfos on WinRTFileSystem.
https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs#L231

This should be implemented by grabbing a StorageFolder for the path, then constructing a query and calling CreateFileQueryWithOptions.  

- Translate SearchOption into [FolderDepth](https://msdn.microsoft.com/en-us/library/windows/apps/windows.storage.search.queryoptions.folderdepth).
- Split the searchPattern into `name` and `extension`
- Set [ApplicationSearchFilter](https://msdn.microsoft.com/en-us/library/windows/apps/windows.storage.search.queryoptions.applicationsearchfilter.aspx) as `$"System.FileName:~\"{name}\""` using [AQS](https://msdn.microsoft.com/en-us/library/windows/apps/bb266512.aspx)
- Set [FileTypeFilter](https://msdn.microsoft.com/en-us/library/windows/apps/windows.storage.search.queryoptions.filetypefilter.aspx) as `extension`, if it exists.

### PR DESCRIPTION
**Pull request by ianhays**
The enumeration in WinRTFileSystem is currently stubbed out to return an empty array. This commit adds an implementation with behavior matching the Win32FileSystem as much as possbible. 

There are still a few differences:
- Search paths that contain a file with an extension containing a wildcard will have those extensions ignored when filtering results.
- Double dot ("..") search patterns do not throw an ArgumentException like they do on Win32.

resolves #8108 

@ericstj

### CODE REVIEW

**Comment by ericstj** [comment]
Can you file issues on the behavior gaps?  What's the reason for not implementing those?

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 1]
```diff
@@ -227,14 +228,86 @@ public override bool DirectoryExists(string fullPath)
...
+            catch (FileNotFoundException e)
+            {
+                // GetFolderFromPathAsync throws FileNotFoundException but win32 throws a DirectoryNotFoundException for this case.
+                throw new DirectoryNotFoundException(e.Message, e.InnerException);
+            }
+
+            // Construct a query for the search.
+            string fileExtension = Path.GetExtension(searchPattern);
+            QueryOptions query;
+            if (fileExtension.Length > 0 && !fileExtension.Contains("?") && !fileExtension.Contains("*")) // The Query won't allow wildcards in file extensions.
```
Comment (ericstj) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: You might be able to use something like `"System.FileExtension:~\"{fileExtension}\""` and leave fileTypes unset.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 2]
```diff
@@ -227,14 +228,86 @@ public override bool DirectoryExists(string fullPath)
...
+            IEnumerable<IStorageItem> storageFiles = SynchronousResultOf(EnumerateFileQuery(fullPath, searchPattern, searchOption, searchTarget));
+
+            // Translate the StorageFile results into FileSystemInfos
+            foreach (IStorageItem item in storageFiles)
+            {
+                if (item.IsOfType(StorageItemTypes.Folder))
+                {
+                    yield return new DirectoryInfo(item.Path);
+                }
+                else // If it is neither a File nor folder then we treat it as a File.
```
Comment (nil4) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Is this comment accurate? There is no test for files above.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 3]
```diff
@@ -227,14 +228,86 @@ public override bool DirectoryExists(string fullPath)
...
+                {
+                    yield return new FileInfo(item.Path);
```
Comment (justinvp) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Potential future perf optimization: The `IStorageItem` could potentially be passed along to the `FileInfo`/`DirectoryInfo` so that it doesn't have to be recreated again when the properties are accessed. WinRT-specific internal constructors would have to be added that would create an instance of [`WinRTFileSystemObject`](https://github.com/dotnet/corefx/blob/d0dc5fc099946adc1035b34a8b1f6042eddb0c75/src/System.IO.FileSystem/src/System/IO/WinRTFileSystemObject.cs#L22) from the `IStoreItem`, and then set the [WinRT FileSystemInfo's `IFileSystemObject`](https://github.com/dotnet/corefx/blob/dad7638e417019f9173f6bbd136bc5a5cd833f4a/src/System.IO.FileSystem/src/System/IO/FileSystemInfo.WinRT.cs#L9) field to that instance when the `FileInfo`/`DirectoryInfo` is created.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 4]
```diff
@@ -227,14 +228,86 @@ public override bool DirectoryExists(string fullPath)
...
+            }
+        }
+
+        private async static Task<IEnumerable<IStorageItem>> EnumerateFileQuery(string path, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
+        {
+            // Get a StorageFolder for "path"
+            StorageFolder folder;
+            try
+            {
+                folder = await StorageFolder.GetFolderFromPathAsync(Path.GetFullPath(path));
```
Comment (justinvp) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Should all of these awaits have `ConfigureAwait(false);`?

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 5]
```diff
@@ -227,14 +228,86 @@ public override bool DirectoryExists(string fullPath)
...
+
+            // Translate SearchOption into FolderDepth
+            FolderDepth depth = searchOption == SearchOption.AllDirectories ? FolderDepth.Deep : FolderDepth.Shallow;
+            query.FolderDepth = depth;
+
+            // Set ApplicationSearchFilter as $"System.FileName:~\"{name}\"" using AQS
+            string fileName = Path.GetFileNameWithoutExtension(searchPattern);
+            if (searchPattern.Equals("."))
+                fileName = "*"; // A period is equivalent to an asterix to indicate a wildcard search.
+            string aqs = string.Format("System.FileName:~\"{0}\"", fileName);
```
Comment (justinvp) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Nit: string concatenation would be a little more efficient here, e.g.

```csharp
string aqs = "System.FileName:~\"" + fileName + "\"";
```

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 6]
```diff
@@ -227,14 +228,86 @@ public override bool DirectoryExists(string fullPath)
...
+            }
+
+            // Construct a query for the search.
+            string fileExtension = Path.GetExtension(searchPattern);
+            QueryOptions query;
+            if (fileExtension.Length > 0 && !fileExtension.Contains("?") && !fileExtension.Contains("*")) // The Query won't allow wildcards in file extensions.
+            { 
+                var fileTypes = new List<string>(1);
+                fileTypes.Add(fileExtension);
+                query = new QueryOptions(CommonFileQuery.DefaultQuery, fileTypes);
```
Comment (justinvp) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Nit: Could this just use an array instead of a `List<T>` to avoid the extra object allocation? e.g.

```csharp
var fileTypes = new string[] { fileExtension };
query = new QueryOptions(CommonFileQuery.DefaultQuery, fileTypes);
```

**Comment by justinvp** [comment]
Originally, I was going to suggest returning `Task<IStorageItem[]>` from the private method, using an array instead of `List<IStorageItem>` and allocating the array based on the size of the resulting `IReadOnlyList<IStorageItem>`s, but why not avoid the array/List<T> allocation altogether, and just enumerate the results directly? Not only would it avoid a (potentially large) array allocation, it also avoids any inefficiencies with `List<T>.AddRange` and resizing the inner array.

Something like:

```csharp
public override IEnumerable<string> EnumeratePaths(string fullPath, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
{
    StorageFolder folder = GetStorageFolder(fullPath);
    QueryOptions query = GetQuery(searchPatten, searchOption);
    
    if ((searchTarget & SearchTarget.Files) > 0)
    {
        IReadOnlyList<IStorageItem> items = GetFilesSync(folder, query);
        int count = items.Count;
        for (int i = 0; i < count; i++)
        {
            yield return items[i].Path;
        }
    }

    if ((searchTarget & SearchTarget.Directories) > 0)
    {
        IReadOnlyList<IStorageItem> items = GetFoldersSync(folder, query);
        int count = items.Count;
        for (int i = 0; i < count; i++)
        {
            yield return items[i].Path;
        }
    }
}
```

And then doing something similar in `EnumerateFileSystemInfos`.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 7]
```diff
@@ -227,14 +228,86 @@ public override bool DirectoryExists(string fullPath)
 
         public override IEnumerable<string> EnumeratePaths(string fullPath, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
         {
-            // Fill in implementation
-            return new string[0];
+            IEnumerable<IStorageItem> storageFiles = SynchronousResultOf(EnumerateFileQuery(fullPath, searchPattern, searchOption, searchTarget));
```
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: I think there's a subtle difference between this implementation and the Win32 and Unix implementations.  Because this is implemented as an iterator, calling this method (same feedback goes for the other methods below) isn't going to actually start enumerating... that won't happen until MoveNext is called.  As such, any errors that result from starting to enumerate won't propagate until the first call to MoveNext rather than propagating out of the synchronous call to Enumerate*.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Yes, they should.  Assuming these are WinRT calls, they should be ```... = await ....AsTask().ConfigureAwait();```

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs (hunk 1 above)
Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: What about using System.ItemPathDisplay and giving a full path like we do in Win32? It works as expected for my tests and we don't have to split up the file name/extension.

Comment (ericstj) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: The [docs](https://msdn.microsoft.com/en-us/library/windows/desktop/bb760776(v=vs.85).aspx) seem to indicate that it may be localized:
> If the item is a file or folder, this property includes the extension in all cases, and is localized if a localized name is available.

I suspect that would be cases like the "documents" folder on a non-english machine.  You may want to test that out to see if your approach would work.  They also mention a `ParsingPath` not sure how that is different.

Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: >They also mention a ParsingPath not sure how that is different.

ParsingPath doesn't work for me where ItemPathDisplay does. According to the ParsingPath docs: `If the item cannot be accessed through the Shell namespace, this value is VT_EMPTY.`

>You may want to test that out to see if your approach would work.

I take it that testing that would require I modify the display language of the machine and not just the keyboard language? Changing the keyboard language doesn't stop the tests from passing.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs (hunk 2 above)
Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Yes it is accurate. It's saying that if a StorageItem isn't a folder then we treat it as a file, regardless of whether it is actually a file.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs (hunk 1 above)
Comment (ericstj) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Yeah you need something more than just keyboard.  OS language and potentially app language.  I have a VM if you'd like to try it out.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs (hunk 4 above)
Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Thanks, I'll make them use TranslateWinRTTask like the other async calls in the file. That includes a ConfigureAwait(false).

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs (hunk 1 above)
Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: The tests passed on your german VM using the Pictures folder as the base test directory. I pushed an update with the code I tested with.

**Comment by ianhays** [comment]
Pushed an update and the behavior is now more closely matched with Win32. The two differences in the OP are no longer present.

I finally found a function that can get *both* files and folders so no need for AddRange. I also modified the AQS query to only use ItemPathDisplay so we don't need to split filename/extension.

I still need to resolve the yield issue.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 8]
```diff
@@ -270,44 +271,99 @@ public override IEnumerable<FileSystemInfo> EnumerateFileSystemInfos(string full
...
+                return await queryResult.GetFoldersAsync().TranslateWinRTTask(folder.Path, isDirectory: true);
+            }
+            else
+            {
+                StorageItemQueryResult queryResult = folder.CreateItemQueryWithOptions(query);
+                return await queryResult.GetItemsAsync().TranslateWinRTTask(folder.Path, isDirectory: true);
+            }
+        }
+
+        private static Tuple<string, string> ConstructAQSQuery(string path, string searchPattern)
```
Comment (ericstj) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Tuple is another allocation and rather unreadable, can you just use `out` here instead or break this into two functions if you can come up with logical separation.
Also, you can avoid two calls to GetFolderFromPathAsync by reordering some of this code?

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 9]
```diff
@@ -227,14 +228,107 @@ public override bool DirectoryExists(string fullPath)
 
         public override IEnumerable<string> EnumeratePaths(string fullPath, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
         {
-            // Fill in implementation
-            return new string[0];
+            IEnumerable<IStorageItem> storageFiles = SynchronousResultOf(EnumerateFileQuery(fullPath, searchPattern, searchOption, searchTarget));
+
+            // Yield the paths of the StorageItems
+            foreach (IStorageItem item in storageFiles)
```
Comment (justinvp) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Could the return type of `EnumerateFileQuery` be `Task<IReadOnlyList<IStorageItem>>` instead of `Task<IEnumerable<IStorageItem>>`?

This would allow indexing into the `IReadOnlyList<T>`, which I suspect is more efficient than using the enumerator  because it would avoid the enumerator allocation and any adapter/RCW allocations for the WinRT projection/interop, and instead of two calls for each item (e.g. `MoveNext()` and `get_Current`) it can just make one call to the indexer (although, I have not measured to confirm). e.g.

```csharp
int count = storageFiles.Count;
for (int i = 0; i < count; i++)
{
    yield return storageFiles[i].Path;
}
```

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs (hunk 8 above)
Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: >Tuple is another allocation and rather unreadable, can you just use out here instead or break this into two functions if you can come up with logical separation.

I can break it into two functions.

>Also, you can avoid two calls to GetFolderFromPathAsync by reordering some of this code?

I thought about that but decided to leave it in to keep argument validation the same as Win32.

Example:
```EnumerateFileQuery("c:\temp\doesntexist", "c:\temp\doesntexist\reallydoesntexist\*")```
If we replace the first folder query with the second sub-query in this case, our behavior changes from throwing a DirectoryNotFoundException to returning an empty enumerable.

We could do Directory.Exists on the original path I suppose, but I don't know if that's any better than GetFolderFromPathAsync. I assumed it wouldn't be.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs (hunk 7 above)
Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Interesting. I'd forgotten that the code before the first yield doesn't occur until enumeration begins.

It looks like Unix and Win32 get around this with their custom enumerable class, but I'd really prefer to avoid that. How about if I move the yielding to a separate method e.g.

```
        public override IEnumerable<string> EnumeratePaths(string fullPath, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
        {
            IEnumerable<IStorageItem> storageFiles = SynchronousResultOf(EnumerateFileQuery(fullPath, searchPattern, searchOption, searchTarget));
            return TranslatePaths(storageFiles);            
        }

        public static IEnumerable<string> TranslatePaths(IEnumerable<IStorageItem> storageFiles)
        {
            // Yield the paths of the StorageItems
            foreach (IStorageItem item in storageFiles)
            {
                yield return item.Path;
            }
        }
```

That way we do the file entry enumeration at initialization and the translation method is the one doing the yielding.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 10]
```diff
@@ -228,41 +228,57 @@ public override bool DirectoryExists(string fullPath)
...
+        /// </summary>
+        private static IEnumerable<FileSystemInfo> TranslateFileSystemInfos(IReadOnlyList<IStorageItem> storageFiles)
+        {
             // Translate the StorageFile results into FileSystemInfos
-            foreach (IStorageItem item in storageFiles)
+            int count = storageFiles.Count;
+            for (int i = 0; i < count; i++)
             {
-                if (item.IsOfType(StorageItemTypes.Folder))
+                if (storageFiles[i].IsOfType(StorageItemTypes.Folder))
```
Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: could keep a local copy to avoid calling the indexer twice. Don't know that it's worthwhile though.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 11]
```diff
@@ -228,41 +228,57 @@ public override bool DirectoryExists(string fullPath)
...
             }
         }
 
-        private async static Task<IEnumerable<IStorageItem>> EnumerateFileQuery(string path, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
+        /// <summary>
+        /// Translates IStorageItems into string paths
+        /// </summary>
+        private static IEnumerable<string> TranslatePaths(IReadOnlyList<IStorageItem> storageFiles)
+        {
+            // Yield the paths of the StorageItems
```
Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: unnecessary comment here and in TranslateFileSystemInfos

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 12]
```diff
@@ -227,14 +228,121 @@ public override bool DirectoryExists(string fullPath)
...
-            // Fill in implementation
-            return new FileSystemInfo[0];
+            IReadOnlyList<IStorageItem> storageFiles = SynchronousResultOf(EnumerateFileQuery(fullPath, searchPattern, searchOption, searchTarget));
+            return TranslateFileSystemInfos(storageFiles);
+        }
+
+        /// <summary>
+        /// Translates IStorageItems into FileSystemInfos
+        /// </summary>
+        private static IEnumerable<FileSystemInfo> TranslateFileSystemInfos(IReadOnlyList<IStorageItem> storageFiles)
```
Comment (justinvp) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Nit: I find the `Translate` prefix a little confusing at first glance as I'm not sure what it's *translating*. How about using the same naming pattern Linq uses for its iterator methods: public methods like `Select` do the up-front argument validation and then call a second private method named `SelectIterator` (suffixed with `Iterator`) which does the lazy yielding. So `EnumerateFileSystemInfosIterator` instead of `TranslateFileSystemInfos`?

Comment (justinvp) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Another nit: instead of naming the parameter `storageFiles`, maybe it should be named `items` since it could contain directories as well as files. (I'd name the temp variable in the calling method the same).

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 13]
```diff
@@ -227,14 +228,121 @@ public override bool DirectoryExists(string fullPath)
...
+        }
+
+        private async static Task<IReadOnlyList<IStorageItem>> EnumerateFileQuery(string path, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
+        {
+            // Get a StorageFolder for "path"
+            string fullPath = Path.GetFullPath(path);
+            StorageFolder folder;
+            try
+            {
+                folder = await StorageFolder.GetFolderFromPathAsync(fullPath).TranslateWinRTTask(fullPath, isDirectory: true);
```
Comment (justinvp) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: @stephentoub can correct me if I am wrong, but I think these `awaits` still need `ConfigureAwait(false)` (even if it is used inside `TranslateWinRTTask`, anywhere you use `await` needs it).

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs (hunk 10 above)
Comment (justinvp) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: I would as it would avoid two interface dispatches per item.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs (hunk 13 above)
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: That's correct.

Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: So `GetFolderFromPath.TranslateWinRtTask.ConfigureAwait`?

We do not do that for any other awaits in this file that use TranslateWinRTTask. I take it we should be?

Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: > So  GetFolderFromPath.TranslateWinRtTask.ConfigureAwait ?

Yes.  Essentially anywhere in the library we're awaiting something, we should await it with ```.ConfigureAwait(false)```.

> We do not do that for any other awaits in this file that use TranslateWinRTTask. I take it we should be?

So, it depends.  The safest answer is "yes, use it everywhere in the library we await something".  There's a more nuanced answer as well, which is that if we know we've been queued to a thread pool thread (e.g. we just did Task.Run and we're inside the delegate or something like that), you could leave of the ConfigureAwait.  ```ConfigureAwait(false)``` basically overrides the default await behavior that looks for the current SynchronizationContext/TaskScheduler and schedules the continuation back to them; instead, with ConfigureAwait(false), it behaves as if both of those returned null/the default.  This avoids unnecessarily scheduling back to the current context, e.g. the UI thread if that's where the method is called from, and that has two benefits.  One, there's a perf win in avoiding that hop.  Two, if someone blocks waiting for the XxAsync method to return, and that Async method internally won't complete until some continuation it's scheduled executes, and if that continuation can only run on the very thread that's blocking waiting for the async operation to complete, you've now deadlocked... using ConfigureAwait helps avoid that.

Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: [TranslateWinRTTask returns the `ConfiguredTaskAwaitable` object directly](https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/WinRTIOExtensions.cs). If we have to ConfigureAwait on the result of that anyways, then the helper method has no purpose and we should be calling `TranslateWinRTTaskCore` from WinRTFileSystem. I'm guessing the Translate method was added as a helper to avoid the need for a ConfigureAwait at every WinRT async call site; if it's not doing that then it should be removed to save us the indirection.

Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Ah!  No, that's fine then.  The important thing is the type of what's being awaited.

**Comment by ianhays** [comment]
Thanks for the feedback! I rebased to one commit.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 14]
```diff
@@ -225,16 +226,122 @@ public override bool DirectoryExists(string fullPath)
...
+                // to make sure something is actually returned when the searchPattern is a subdirectory of the path.
+                // To do this, we attempt to get a new StorageFolder for the subdirectory and return an empty enumerable
+                // if we can't.
+                string searchPatternDirName = Path.GetDirectoryName(normalizedSearchPattern);
+                string userPath = string.IsNullOrEmpty(searchPatternDirName) ? fullPath : Path.Combine(fullPath, searchPatternDirName);
+                if (userPath != folder.Path)
+                {
+                    try
+                    {
+                        folder = await StorageFolder.GetFolderFromPathAsync(userPath).TranslateWinRTTask(userPath, isDirectory: true);
```
Comment (ericstj) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: You're still doing an extra call to GetFolderFromPathAsync here.  You don't use folder anywhere between it's initial creation and here.  You could combine this call and the previous to avoid throw-away work.

Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: That is by design. The first call asserts that the initial folder exists and throws a DirectoryNotFoundException if it doesn't. In the case of a missing watcher path, it is not throw away work.

Cases I test:

path | searchPattern | result
------------ | ----------------- | --------------------------
```c:\temp\doesntexist``` | ```c:\temp\doesntexist\reallydoesntexist\*``` | throws DirectoryNotFoundException on first query
```c:\temp\doesntexist``` | ```c:\temp\doesntexist\*``` | throws DirectoryNotFoundException on first query
```c:\temp\exists``` | ```c:\temp\exists\subdirexists\*``` | does two queries - returns actual results
```c:\temp\exists``` | ```c:\temp\exists\subdirdoesntexist\*``` | does two queries - returns empty array

We can consolidate the calls, but it will require that we either:
- diverge from the win32 behavior and not throw a directorynotfoundexception on the first case in the table above
- Do a Directory.Exists call on the path when we have a subdirectory filter.

I don't want to diverge behavior (even for this edge case), so we need to do some sort of directory.exists call. The reason I left the second GetFolderFromPathAsync call is because it seemed cleaner and more reliable than adding a call to API that may determine existence of a directory differently (e.g. Directory.Exists may handle links differently than GetFolderFromPathAsync). Then there's the question of whether Directory.Exists is even *that much* faster than GetFolderFromPathAsync. 

There's also the fact that for the invalid-subpath case (number 1 in the table above), we currently quit execution on the first query and avoid doing path normalization and validation; if we consolidated the calls then we would then we would need to do two calls (GetFolderFromPathAsync and DirectoryExists) whereas now we only need one. We'd also be  slowing down **all** cases where the path doesn't exist because we'd be doing unnecessary path normalization before getting the (missing/invalid) folder.

TL;DR: the duplicate call is for checking the existence of the path before including the searchPattern and normalizing.
.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs [hunk 15]
```diff
@@ -225,16 +226,122 @@ public override bool DirectoryExists(string fullPath)
...
+            string fullPath = Path.GetFullPath(path);
+            StorageFolder folder;
+            try
+            {
+                folder = await StorageFolder.GetFolderFromPathAsync(fullPath).TranslateWinRTTask(fullPath, isDirectory: true);
+            }
+            catch (FileNotFoundException e)
+            {
+                // GetFolderFromPathAsync throws FileNotFoundException but win32 throws a DirectoryNotFoundException for this case.
+                throw new DirectoryNotFoundException(e.Message, e.InnerException);
```
Comment (ericstj) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Shouldn't this be handled by TranslateWinRTTask?

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs (hunk 14 above)
Comment (ericstj) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: I see I missed that we were intentionally doing the first to get the exception.

File: src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs (hunk 15 above)
Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs]: Ah, so it is. Removed.

**Comment by ianhays** [comment]
test Innerloop OSX Release Build and Test  Innerloop OSX Release Build and Test please

**Comment by ianhays** [comment]
test Innerloop OSX Release Build and Test please

**Comment by ianhays** [comment]
@ericstj any more feedback before I rebase again and merge?

**Comment by ericstj** [comment]
LGTM, go ahead
